sandbox/gpio/common.go

```diff
@@ -0,0 +1,315 @@
+		_ = flock.Close()
+	}, nil
+}
+
+var aggregatorCreationTimeout = 120 * time.Second
+
+func addAggregatedChip(ctx context.Context, sourceChip *ChardevChip, lines strutil.Range) (chip *ChardevChip, err error) {
+	// synchronize gpio helpers' access to the aggregator interface
+	unlocker, err := lockAggregator()
+	if err != nil {
+		return nil, err
+	}
+	defer unlocker()
+
+	watcher, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	defer watcher.Close()
+
+	err = watcher.AddWatch(dirs.DevDir, inotify.InCreate)
+	if err != nil {
+		return nil, err
+	}
+
+	// <label> <lines>
+	cmd := fmt.Sprintf("%s %s", sourceChip.Label, lines.String())
+	if err := os.WriteFile(filepath.Join(dirs.GlobalRootDir, aggregatorNewDevicePath), []byte(cmd), 0644); err != nil {
+		return nil, err
+	}
+
+	timeoutC := time.After(aggregatorCreationTimeout)
+	for {
+		select {
+		case event := <-watcher.Event:
+			path := event.Name
+			// check prefix /dev/gpiochipX
+			if strings.HasPrefix(path, filepath.Join(dirs.DevDir, "gpiochip")) {
+				return chardevChipInfo(path)
+			}
+		case <-timeoutC:
+			return nil, errors.New("timeout waiting for aggregator device to appear")
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+func aggregatedChipUdevRulePath(instanceName, slotName string) string {
+	fname := fmt.Sprintf("69-snap.%s.interface.gpio-chardev-%s.rules", instanceName, slotName)
+	return filepath.Join(filepath.Join(dirs.GlobalRootDir, ephemeralUdevRulesDir), fname)
+}
+
+func addEphemeralUdevTaggingRule(ctx context.Context, chip *ChardevChip, instanceName, slotName string) error {
+	if err := os.MkdirAll(filepath.Join(dirs.GlobalRootDir, ephemeralUdevRulesDir), 0755); err != nil {
+		return err
+	}
+
+	tag := fmt.Sprintf("snap_%s_interface_gpio_chardev_%s", instanceName, slotName)
+	rule := fmt.Sprintf("SUBSYSTEM==\"gpio\", KERNEL==\"%s\", TAG+=\"%s\"\n", chip.Name, tag)
+
+	path := aggregatedChipUdevRulePath(instanceName, slotName)
+	if err := os.WriteFile(path, []byte(rule), 0644); err != nil {
+		return err
+	}
+
+	// make sure the rule we just dropped is loaded as sometimes it doesn't get
+	// picked up right away
+	output, err := exec.CommandContext(ctx, "udevadm", "control", "--reload-rules").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cannot reload udev rules: %w", osutil.OutputErr(output, err))
+	}
+	// trigger the tagging rule
+	output, err = exec.CommandContext(ctx, "udevadm", "trigger", "--name-match", chip.Name).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cannot trigger udev rules: %w", osutil.OutputErr(output, err))
+	}
+
+	return nil
+}
+
+var osStat = os.Stat
+var osChmod = os.Chmod
+var osChown = os.Chown
+var syscallMknod = syscall.Mknod
+
+func addGadgetSlotDevice(chip *ChardevChip, instanceName, slotName string) error {
+	finfo, err := osStat(chip.Path)
+	if err != nil {
+		return err
+	}
+
+	stat, ok := finfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return errors.New("internal error")
```

Comment:
Can you perhaps make the error more discoverable (unique):
```suggestion
		return errors.New("internal error, expected os.File.FileInfo.Sys to return *syscall.Stat_t")
```